### ISSUE TITLE
Collection initialisation

### ISSUE THREAD

**Issue by pushbit** [immutables/immutables#307]
Currently `Collection` fields are initialised with a new instance of a concrete type.  In the case where zero elements are added to the collection, the allocation is unnecessary.  What do you think about initialising `Collection` fields with a static empty instance (e.g. `Collections.EMPTY_LIST`) and then only instantiating a concrete class when the first element is added?

I have classes that have ten or more `Collection` fields and which fields will actually be populated in each instance is highly variable.  When tens or hundreds of these objects are instantiated (along with all of their `Collection` fields) it is certainly less than ideal in environments, such as mobile, where object allocation and garbage collection are relatively expensive.

**Comment by elucash** [comment]
Thank you for feature request! It was one of those early small design decision where we made some micro-benchmarking and figured out that it's more cpu efficient and memory negligible to just create a builder rather than create it on-demand when first element need to be added. Of course this story was not about mobile/Android, it's even about using Guava builders, not the Jdk-only construction paths implemented later. For the JDK-only [collection] construction path there are already some places where we doing additional checks to maintain compactness/correctness of immutable/unmodifiable collections.
I'll try to revise some parts of construction to see how easy/difficult would be to implement that.

**Comment by elucash** [comment]
2.2-beta released with revised building of jdk collections (list, sets, map, except for enum based ones), inviting to evaluate it

**Comment by pushbit** [comment]
It's working nicely with `@Immutable` + `@Style(jdkOnly = true)`.  Would it be possible to also implement for Guava collections and `@Modifiable`?

**Comment by elucash** [comment]
Revised implementation to better align with #299  . `null` is used until actual value is needed. As referencing `Collections.emptyList` and friends does not involve allocation, this should work ok. JDK, Guava collections and Modifiable objects are now supported unless `strictBuilder = true` is used to prevent this optimization.

Released in 2.2-beta-2

**Comment by pushbit** [comment]
`@Immutable` looks good.  `@Modifiable` looks close, but I didn't expect that a collection accessor would instantiate a new instance if the field was empty/null.  I guess this is so the collection could be modified externally, like would be the case if the collection was already populated.

Unfortunately this means that a new instance will be allocated even if the accessor is only used to read the contents of an empty collection.  Perhaps an `isFooEmpty()` method per collection field would be a reasonable addition?  (Better ideas welcome. :))  It wouldn't be obvious on first glance, but at least there would be some way for clients to know that a collection field is empty without having to instantiate it.

**Comment by elucash** [comment]
> I guess this is so the collection could be modified externally

exactly

> Unfortunately this means that a new instance will be allocated even if the accessor is only used to read the contents of an empty collection.

yes, but I've made another optimization to allocate with 0 internal capacity

> Perhaps an isFooEmpty() method per collection field would be a reasonable addition?

I thought about this, but this is an optimization that should not be observed externally. We have `IsSet` methods only for mandatory/default attributes.

Another solution comes to mind is that you can have collection with default value (in 2.2)

```java
@Value.Default
public List<String> getList() {
  return Collections.emptyList();
}
```

Unfortunately I haven't tried how well it works for Modifiable objects.

Please, try 2.2-beta-3 with fixes and refinements

**Comment by pushbit** [comment]
The optimization to allocate with 0 internal capacity may be good for the accessor when the field is `null`, but this is also being done in the `add` and `set` methods, which I assume is unintended.

I added a `@Default` and it does appear that this would work for returning an `empty*()` when the field is `null`, though I'm not sure this is a good idea because clients may be surprised to get an unmodifiable collection when it's empty.

But the bigger problem is that the generated `@Modifiable` doesn't compile because there are conflicting `*IsSet()` methods.  It seems that we only want the latter because `optBits` is never set when the collection is modified.  Would there be any harm in including the latter `*IsSet()` even when a `@Default` is not specified?  This would allow clients to avoid the allocation in the accessor when the field is `null`.

```java
  public final boolean barIsSet() {
    return (optBits & OPT_BIT_BAR) != 0;
  }

  public final boolean barIsSet() {
    return bar != null;
  }
```

**Comment by elucash** [comment]
Made another revision. No new version published as this is the only change since beta-3, but you can take snapshots `2.2-beta-4-SNAPSHOT` from oss sonatype repo as shown in #137, or build from source.

I think I failed to introduce this as an invisible optimization, so resorted to use style flag `deferCollectionAllocation = true`, this will also gives a method on Modifiable to check if `fooIsSet()` for collections as you suggested. It took some time to sort this out better and still cannot guarantee full satisfaction :-) , but it helped to nail down a lot of other issues which were not previously apparent.

**Comment by pushbit** [comment]
I think it looks good. :)

But then I switched the collections to arrays and there's a few problems.

```java
	@Immutable
	abstract class Foo {
		@Default
		int[] foo() {
			return ArrayUtils.EMPTY_INT_ARRAY;
		}
	}
```
generates
```java
  private ImmutableFoo(ImmutableFoo.Builder builder) {
    this.foo = builder.foo != null
        ? builder.foo
        : foo.clone();
  }

  @Override
  int[] foo() {
    return foo.clone();
  }
```

Last line of the constructor causes a compile error because `foo` isn't initialised and the accessor doesn't reference the `@Default` super method.

**Comment by elucash** [comment]
That was really helpful, snapshot contains the fix

**Comment by elucash** [comment]
released in 2.2 Thank you!
